src/main/java/org/protege/owlapi/concurrent/WriteSafeOWLOntologyImpl.java

```diff
@@ -1318,4 +1297,861 @@ public boolean isTopEntity() {
         }
     } 
     
+    @Override
+    public void accept(OWLNamedObjectVisitor visitor) {
+        readLock.lock();
+        try {
+            delegate.accept(visitor);
+        } finally {
+            readLock.unlock();
+        }
+    }
+
+    @Override
+    public <O> O accept(OWLNamedObjectVisitorEx<O> visitor) {
+        readLock.lock();
+        try {
+            return delegate.accept(visitor);
+        } finally {
+            readLock.unlock();
+        }
+    }
+
+    @Override
+    public void setOWLOntologyManager(OWLOntologyManager saveOntology) {
+        writeLock.lock();
+        try {
+            delegate.setOWLOntologyManager(saveOntology);
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    @Override
+    public void saveOntology() throws OWLOntologyStorageException {
+        writeLock.lock();
```

Comment:
> If it's expedient, we can do it for 4.0.2. 

Then 4.0.2 should become 4.1.0 :)  
My feeling is that doing this would help us a lot with maintainability.

> What would be really useful is some test ensuring the machinery operates as intended. These sort of tests are hard to come by though - any chance of any such thing existing already?

It depends what you mean by operates as intended.  I have some *simple* unit tests like the one below that ensures that the read lock is locked and unlocked in the correct order, but that's it (ontology is a mock btw).

```java
    @Test
    public void shouldDelegateTo_getSameIndividualAxioms_withReadLock() throws Exception {
        OWLIndividual arg0 = mock(OWLIndividual.class);

        ontology.getSameIndividualAxioms(arg0);

        InOrder inOrder = Mockito.inOrder(readLock, delegate, readLock);
        inOrder.verify(readLock, times(1)).lock();
        inOrder.verify(delegate, times(1)).getSameIndividualAxioms(arg0);
        inOrder.verify(readLock, times(1)).unlock();

        verify(writeLock, never()).lock();
        verify(writeLock, never()).unlock();
    }
```